lib/pacemaker/pcmk_sched_colocation.c

```diff
@@ -1337,19 +1337,36 @@ add_node_scores_matching_attr(GHashTable *nodes, const pe_resource_t *rsc,
         int weight = 0;
         int score = 0;
         int new_score = 0;
+        const char *value = pe_node_attribute_raw(node,
+                                                  colocation->node_attribute);
 
-        score = best_node_score_matching_attr(rsc, attr,
-                                              pe_node_attribute_raw(node, attr));
+        score = best_node_score_matching_attr(rsc, colocation->node_attribute,
+                                              value);
 
         if ((factor < 0) && (score < 0)) {
-            /* Negative preference for a node with a negative score
-             * should not become a positive preference.
+            /* If the dependent is anti-colocated, we generally don't want the
+             * primary to prefer nodes that the dependent avoids. That could
+             * lead to unnecessary shuffling of the primary when the dependent
+             * hits its migration threshold somewhere, for example.
              *
-             * @TODO Consider filtering only if weight is -INFINITY
+             * However, there are cases where it is desirable. If the dependent
+             * can't run anywhere but where the primary is, it would be
+             * worthwhile to move the primary for the sake of keeping the
+             * dependent active.
+             *
+             * We use the primary's stickiness to decide what to do, so the user
+             * has some say in the matter. (This is the configured primary, not
+             * a particular instance of the primary, but that doesn't matter
+             * unless stickiness uses a rule to vary by node, and that seems
+             * acceptable to ignore.)
              */
-            crm_trace("%s: Filtering %d + %f * %d (double negative disallowed)",
-                      pe__node_name(node), node->weight, factor, score);
-            continue;
+            if ((colocation == NULL)
+                || (colocation->primary->stickiness >= -score)) {
```

Comment:
This approach is still overzealous IMO but it's acceptable. Calling back to https://github.com/ClusterLabs/pacemaker/pull/3092#discussion_r1201292245:

> [Suppose a situation where]
> 
> * R1 is running on node1
> * R2 has a negative preference for node2
> * R2 has no preference regarding node1 or node3
> 
> Even with no stickiness, there's no logical reason to move R1 to node2 in this scenario. The only reason I see would be convenience of implementation, using "well, there's no stickiness..." as a rationalization for why it's okay to do so.
> 
> You could make the case for that, so that we can get it done and move on without more dramatic code changes. The behavior doesn't seem ideal though.

Emphasis on "Even with no stickiness". With this implementation, if I read it correctly, R1 would move to node2 in the above scenario, even though R2 could just as well run on node3 without disrupting R1.

This becomes especially relevant when `score` is `-INFINITY` due to, say, the dependent hitting its migration threshold. Usually we don't configure `resource-stickiness=INFINITY`. Infinite stickiness is often more extreme behavior than what we want, and it probably shouldn't be required just to prevent the primary from shuffling in this scenario.

The new tests don't cover this scenario.

Reply:
Hmm, I'll have to think about that

Updated commits for other comments

Reply:
Updated to consider additional factors